src/ws/session.c

```diff
@@ -562,16 +562,57 @@ perform_tlscert (void)
 }
 
 
+static void
+store_krb_credentials (gss_cred_id_t creds, uid_t uid, gid_t gid)
+{
+  assert (creds != GSS_C_NO_CREDENTIAL);
+
+  bool was_root = getuid() == 0;
+
+  /* We want to do this before the fork(), so we need to temporarily
+   * change our euid/egid
+   */
+  if (setresgid (gid, gid, -1) != 0 || setresuid (uid, uid, -1) != 0)
+    err (127, "Unable to temporarily drop permissions to store gss credentials");
+
+  assert (geteuid () == uid && getegid() == gid);
+
+  krb5_context k5;
+  int res = krb5_init_context (&k5);
```

Comment:
Is this documented safe for not being able to run/inject arbitrary user-controlled code during the negotiation? This approach now puts the entire weight of security into the kerberos library, whereas before everything already ran as the target user. "ssh does it" alleviates my concern a bit, but I'd like to have at least a good justification about what advantage this has.

Please don't call the previous attempt "gross" -- *this* approach is. It makes a *lot* more sense to do a permanent privilege changes after fork than having to do this brittle "temporarily change uid before and hope that you didn't leave a loophole". There may be a reason for doing it that way, but that needs a better rationale than an emotion. :-)